05-onchain.md

```diff
@@ -557,6 +569,18 @@ Thus, 483 bidirectional HTLCs (containing both `to_local` and
 Note: even if the `to_remote` output is not swept, the resulting
 `max_num_htlcs` is 967; which yields the same unidirectional limit of 483 HTLCs.
 
+# Generation of HTLC Transactions
+
+If `option_anchor_outputs` does not apply to the commitment transaction, then HTLC-timeout and HTLC-success transactions are complete transactions with (hopefully!) reasonable fees and must be used directly.
+
+Otherwise, the use of `SIGHASH_SINGLE|SIGHASH_ANYONECANPAY` on the HTLC signatures received from the peer allows HTLC transactions to be combined with other transactions.
+
+## Requirements
+
+A node which broadcasts an HTLC-success or HTLC-timeout transaction for a commitment transaction for which `option_anchor_outputs` applies:
+  - MUST contribute sufficient fee to ensure timely inclusion in a block.
+  - MAY combine it with other transactions.
```

Comment:
nit: you can precise "MAY combine it with other non-HTLC-timeout/HTLC-success" as all of them are going to pre-sign the same index value

Reply:
Not sure what you mean by this. Every htlc has its own output index on the commitment transaction?

Reply:
You can't combine multiple HTLC-success or HTLC-timeout txn as they are all going to sign the input index and output index 0. If you're transaction is dual or multiparty signed you can't aggregate freely without new interactivity.

Reply:
I didn't realize that. That means that an additional fee input needs to be added to every htlc tx, assuming there is not much else to batch it with? Needs many utxos if there are 400 something htlcs outstanding :(

Reply:
I am no expert here, but I took a look at the sighash algorithm, and it seems like this should work with segwit, since it no longer commits to the index: 

> SINGLE does not commit to the input index. When ANYONECANPAY is not set, the semantics are unchanged since hashPrevouts and outpoint together implictly commit to the input index. When SINGLE is used with ANYONECANPAY, omission of the index commitment allows permutation of the input-output pairs, as long as each pair is located at an equivalent index.

https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#specification

Reply:
Ah you're right thanks, it wouldn't have been possible before SegWit and the new transaction digest algorithm but that's cool and should save us a lot of utxos!

Here the lines of concern : https://github.com/bitcoin/bitcoin/blob/e65b4160e99fa86d6d840dce75cae29334afd1f2/src/script/interpreter.cpp#L1257

So we can aggregate txn but I think we should be careful in case of RBF bump, like not adding new unconfirmed input to avoid choking against BIP 125 rules 2 (but I think we can drop a input-output pair and broadcast on its own if its timelock expires soon ?)

Reply:
The way this currently works in `lnd` is that we use exponential randomized back-off per input if a constructed transaction fails to publish. We have no insight into the mempool (neutrino) and error messages don't allow us to pinpoint exactly what is wrong. It isn't ideal and a whole area to improve upon further.

Reply:
Okay digging on this, and reading segwit v0 transaction digest code, as raised by @halseth aggregation isn't an issue on the sighash level, you can freely combine input-ouput pair. However, aggregation of HTLC-timeout with different locktime will break signatures as `nLocktime` is committed. So should be amended to `MAY be combined with other HTLC-input-output comitting to the same nLocktime value` (I would avoid using transaction here as we move to a per-input-ouput _granularity_)